path/path.go

```diff
@@ -0,0 +1,129 @@
+	// against the path should protect against modification of the original.
+	steps PathSteps
+}
+
+// AtListIndex returns a copied path with a new list index step at the end.
+// The returned path is safe to modify without affecting the original.
+func (p Path) AtListIndex(index int) Path {
+	copiedPath := p.Copy()
+
+	copiedPath.steps.Append(PathStepElementKeyInt(index))
+
+	return copiedPath
+}
+
+// AtMapKey returns a copied path with a new map key step at the end.
+// The returned path is safe to modify without affecting the original.
+func (p Path) AtMapKey(key string) Path {
+	copiedPath := p.Copy()
+
+	copiedPath.steps.Append(PathStepElementKeyString(key))
+
+	return copiedPath
+}
+
+// AtName returns a copied path with a new attribute or block name step at the
+// end. The returned path is safe to modify without affecting the original.
+func (p Path) AtName(name string) Path {
+	copiedPath := p.Copy()
+
+	copiedPath.steps.Append(PathStepAttributeName(name))
+
+	return copiedPath
+}
+
+// AtSetValue returns a copied path with a new set value step at the end.
+// The returned path is safe to modify without affecting the original.
+func (p Path) AtSetValue(value attr.Value) Path {
+	copiedPath := p.Copy()
+
+	copiedPath.steps.Append(PathStepElementKeyValue{Value: value})
+
+	return copiedPath
+}
+
+// Copy returns a duplicate of the path that is safe to modify without
+// affecting the original.
+func (p Path) Copy() Path {
+	return Path{
+		steps: p.Steps(),
+	}
+}
+
+// Equal returns true if the given path is exactly equivalent.
+func (p Path) Equal(o Path) bool {
+	if p.steps == nil && o.steps == nil {
+		return true
+	}
+
+	if p.steps == nil {
+		return false
+	}
+
+	if !p.steps.Equal(o.steps) {
+		return false
+	}
+
+	return true
+}
+
+// ParentPath returns a copy of the path with the last step removed.
+//
+// If the current path is empty, an empty path is returned.
+func (p Path) ParentPath() Path {
+	if len(p.steps) == 0 {
+		return EmptyPath()
+	}
+
+	_, remainingSteps := p.steps.Copy().LastStep()
+
+	return Path{
+		steps: remainingSteps,
+	}
+}
+
+// Steps returns a copy of the underlying path steps. Returns an empty
+// collection of steps if path is nil.
+func (p Path) Steps() PathSteps {
+	if len(p.steps) == 0 {
+		return PathSteps{}
+	}
+
+	return p.steps.Copy()
+}
+
+// String returns the human-readable representation of the path.
+// It is intended for logging and error messages and is not protected by
+// compatibility guarantees.
+func (p Path) String() string {
+	return p.steps.String()
+}
+
+// EmptyPath creates an empty attribute path. Provider code should use
+// RootPath.
+func EmptyPath() Path {
+	return Path{
+		steps: PathSteps{},
+	}
+}
+
+// RootPath creates a new attribute path starting with a root
+// PathStepAttributeName.
+func RootPath(rootAttributeName string) Path {
+	return Path{
+		steps: PathSteps{
+			PathStepAttributeName(rootAttributeName),
+		},
+	}
+}
```

Comment:
Given it worked out that this could be in its own package, I wonder if its worth shortening these these to just `path.Root(string)` and `path.Empty()`?

Reply:
I'll update and push up, one second